src/AuxFunctionAlgorithm.C

```diff
@@ -78,6 +79,9 @@ AuxFunctionAlgorithm::execute()
 
     auxFunction_->evaluate(coords, time, nDim, length, fieldData, fieldSize);
   }
+
+  field_->modify_on_host();
+  field_->sync_to_device();
```

Comment:
Are you sure we want the sync_to_device here? It seems like it should be located in/near the code that needs the field on device.

Reply:
The only thing is performance penalty right? I feel that we should get it correct first and then profile and fix performance. 

Also we call a lot of the ngp algorithms through things like `Assemble*SolverAlgorithm`. As written, it is not possible to know which fields are going to be accessed. If we change the interfaces such that we can call `sync_to_device` on ElemDataRequests (but this doesn't exist for edge and node algorithms) then we can do what you are suggesting.